gr-fec/lib/awgn_bp.cc

```diff
@@ -99,39 +105,75 @@ void awgn_bp::spa_initialize()
 
 void awgn_bp::update_chks()
 {
-    double product, _prdct;
+    double sign_prod, tanh_prod, atanh, x;
     int v;
+    int w;
+    int sign;
+    //(step 1) of LLR-BP tanh algo.
+    //iterate over rows of check matrix
     for (int chk = 0; chk < M; chk++) {
-        product = double(1.0);
+        //iterate over nonzero cols of check matrix
         for (int i = 0; i < num_mlist[chk]; i++) {
+            sign_prod = double(1.0);
+            tanh_prod = double(1.0);
             v = mlist[chk][i] - 1;
-            product = product * double(2 / (1 + Q[chk][v]) - 1);
-        }
-        for (int i = 0; i < num_mlist[chk]; i++) {
-            v = mlist[chk][i] - 1;
-            _prdct = product / double(2 / (1 + Q[chk][v]) - 1);
-            R[chk][v] = double((1 - _prdct) / (1 + _prdct));
+            //compute product of sign and and tanh LLR's
+            for (int iprime = 0; iprime < num_mlist[chk]; iprime++){
+                //exclude iprime from the product set
+                if (iprime == i){iprime++;}
+                if (iprime >= num_mlist[chk]){break;}
+                w = mlist[chk][iprime] - 1;
+                //compute prod(sign(LLR))
+                if (std::signbit(Q[chk][w])){sign=-1;}
+                else {sign=1;}
+                sign_prod = sign_prod * sign;
+                //remove divide by 2 for manual tanh
+                x = std::abs(Q[chk][w])/2.0;
+                //x = std::abs(Q[chk][w]);
+                //clamp tanh input (this is some BS, FIXME)
+                if (x > 2.35){x = 2.35;}
```

Comment:
I just realized that the code is using `double`'s rather than `float`'s. Still I think that my suggestion to use the `float32` machine epsilon is good because the code probably would also work equally well (in terms of sensitivity) with `float`'s instead of `double`'s, because the LLR calculations can be successfully carrier out in integer arithmetic with just a few bits of quantization.

Reply:
Okay this is tricky. I have experimented with this for a while. 2.35 specifically gives the best results. Not more not less. I had the exact same idea with the epsilon. I think it makes sense. I have no idea why it gives inferior results. The clamp did have to be in the internal loop though because that is where the first NaN or Inf values would appear. I'm really not sure wtf is going on with this, and I have not yet figured it out.

Reply:
> where the function `phi` is defined as phi(x)=log⁡ex+1ex−1, and can be computed with a lookup table. 

As far as your first comment on the algo. This was a very interesting find. I went ahead and tried it and it seemed to work just as well as the current code save for a bit of a CPU performance hit due to the extra logs and exponents. For the most part... Turns out, it suffers from the same zero problem that the original algo did no doubt because of the nested log(exp()) precision issues. I tried clamping the **phi** function values too, but I wasn't immediately successful. I imagine that having **phi** as an actual LUT though would probably solve both the zero and performance issues. For now I have reverted to what is working more reliably since I just didn't have time to troubleshoot where the NaNs were coming from this time.

> It also mentions that the product of signs can be carried out using integers and modulo 2 arithmetic, which is another performance improvement that we can apply to this code.

I also investigated this briefly as it was mentioned in the original reference I used as well. This makes great sense, but just turned out to harder than I'd hoped to implement. Will have to revisit this.

Reply:
> The clamp did have to be in the internal loop though because that is where the first NaN or Inf values would appear. I'm really not sure wtf is going on with this, and I have not yet figured it out.

Am I understanding correctly that `tanh()` is the function that is making the NaN's or Inf's appear in the first place? I'm testing what happens if we evaluate the `tanh()` of a large number, such as `100.0`, and it gives `1.0`: https://godbolt.org/z/TWrsda9bf

Reply:
That was indeed the case. But let me check what happens now.

Reply:
if I clamp the value in the outer loop (tanh_prod=0.982 or tanh_prod=1-1e-7) I get bursts of bad BER. I can't explain it.

Reply:
Okay. I don't understand why that happens either, but that's just a problem with my understanding. The code never lies, so there's something at play we're not noticing.

Let's keep the clamping in the inner loop unless someone can figure out how to do it in the outer loop without any BER performance problems.